src/mlpack/methods/ann/layer/sampling.hpp

```diff
@@ -0,0 +1,168 @@
+  Sampling(const size_t inSize, const size_t outSize);
+
+  /**
+   * Create the Sampling layer object using the specified sample vector size.
+   *
+   * @param layerSize The number of output units.
+   */
+  Sampling(const size_t sampleSize);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename eT>
+  void Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards trough f. Using the results from the feed
+   * forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& input,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating
+   * Kullback–Leibler divergence between a normal distribution
+   * and the standard normal.
+   *
+   * @param input Input data used for evaluating the specified function.
+   */
+  template<typename InputType>
+  double klForward(const InputType&& input);
+
+  /**
+   * Ordinary feed backward pass of a neural network, evaluating the backward
+   * pass of Kullback–Leibler divergence. Using the results from the
+   * KL divergence feed forward pass.
+   *
+   * @param input The propagated input activation.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename InputType, typename OutputType>
+  void klBackward(const InputType&& input,
+                OutputType&& output);
+
+  //! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }
+
+  //! Get the output parameter.
+  OutputDataType const& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the input size.
+  size_t const& InputSize() const { return inSize; }
+  //! Modify the input size.
+  size_t& InputSize() { return inSize; }
+
+  //! Get the output size.
+  size_t const& OutputSize() const { return outSize; }
+  //! Modify the output size.
+  size_t& OutputSize() { return outSize; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored number of input units.
```

Comment:
Here we could just store 'latentSize' to reduce redundancy. 'inSize' would be 2 * latentSize and 'outSize' will be latentSize